Enhanced AindMetadataServiceClient with New Endpoint Methods

dougollerenshaw opened the pull request:
This PR enhances the `AindMetadataServiceClient` class by adding methods for all available API endpoints provided by the metadata service. It also includes comprehensive test coverage and documentation updates.

## Changes

### Client Enhancements
- Added new client methods to access additional endpoints:
  - `get_intended_measurements`: Retrieves intended measurements for a subject
  - `get_injection_materials`: Retrieves injection materials by preparation lot number
  - `get_ecephys_sessions`: Retrieves ecephys sessions for a subject
  - `get_protocols`: Retrieves protocol information by name
  - `get_mgi_allele`: Retrieves MGI allele information
  - `get_perfusions`: Retrieves perfusion information for a subject
  - `get_funding`: Retrieves funding information for a project
  - `get_project_names`: Retrieves all project names
  - `get_smartspim_imaging`: Retrieves SmartSPIM imaging data with optional filtering
  - `get_histology`: Retrieves histology data with optional filtering

### Environment Handling
- Simplified the client constructor to support both production and development endpoints:
  - Added `PROD_URL` and `DEV_URL` class constants
  - Added `use_dev` parameter to toggle between environments
  - Maintained `custom_domain` parameter for custom endpoints

### Documentation
- Updated `README.md` with examples for all new client methods
- Improved installation instructions to prevent shell-specific errors:
  - Added proper quoting for package extras in different shells (bash, zsh, cmd, PowerShell)
  - Updated development installation command with proper quoting

### Testing
- Added comprehensive tests for the new client methods following the existing test pattern
- Fixed issues with the constructor parameter in existing tests

## Benefits
- Complete API coverage: Client now supports all endpoints available in the server
- Improved documentation: Clear examples make it easier for users to understand available functionality
- Better shell compatibility: Fixed installation issues in shells like zsh that interpret square brackets as special characters
- Simplified environment selection: Users can now easily switch between production and development environments

This PR completes the client implementation, ensuring that all available server endpoints have corresponding client methods.

Closes #367

jtyoung84 commented on src/aind_metadata_service/client.py:
```diff
@@ -23,19 +24,43 @@ class StatusCodes(Enum):
 
-    def __init__(self, domain: str):
+    PROD_URL = "http://aind-metadata-service"
+    DEV_URL = "http://aind-metadata-service-dev"
```
Most clients require the user to set the domain (and other credentials) instead of hard-coding them in code.

dougollerenshaw replied:
I'd argue that since most users will be internal that the priority should be ease of use for those users. This pattern avoids the need for users to remember and type the two URLs. It also makes it easy to change the URLs in one place if they need to change in the future (e.g. if we want to start using https instead of http).

For improved modularity, I moved all constants to a new constants.py. Is that an improvement?

dougollerenshaw commented:
Moving Teams conversation between @jtyoung84 and I here for visibility:

> Jon Young Yesterday 6:16 PM
> It's slightly unusual to hardcode domains and hostnames in code. The paths are fine, but we should probably expect the user to set the correct domain.

> Doug Ollerenshaw Yesterday 6:36 PM
> Thanks for exploring the intended_measurements issue!
>  
> For the URLs, I'd really argue that the priority should be ease of use for internal users. Are there any downsides to including the URLs in constants, other than convention? Do you see this repo being used often by external users?
>  
> I'll defer to you in the end if you feel strongly about this. But to me this seems to be a net improvement for usability and maintainability.
>  
> Might be worth getting some input from others too

 

> Jon Young 9:09 AM
> We can open a feature request to read from a local config file stored on the user's computer so they don't have to input the domain name each time. Even if, at the moment, this is only expected to be a internal facing, it should be written following best practices for containerized services. The other stuff you added seemed like a good idea. Maybe we can separate the hardcoded URLs until after more discussion?

dougollerenshaw commented:
@jtyoung84 I just re-requested review. I removed the hard-coded paths and removed `constants.py` altogether. But I kept the paths in the README and the docstring for the function. I also throw a custom error if the user fails to speciify the paths like this:
```
>>> from aind_metadata_service.client import AindMetadataServiceClient
>>> client = AindMetadataServiceClient()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/doug.ollerenshaw/code/aind-metadata-service/src/aind_metadata_service/client.py", line 43, in __init__
    raise ValueError(
ValueError: You must specify the server domain. If you're onsite at the Allen Institute, the production domain is http://aind-metadata-service and the development domain is http://aind-metadata-service-dev
```
This way it will be obvious to the user what is going wrong if they fail to pass the URL and they won't have to go searching around to figure out what the correct URL is. 

Hopefully this addresses your concern while still maintaining maximum ease of use.

jtyoung84 commented on tests/test_client.py:
```diff
@@ -24,6 +24,17 @@
+        self.assertEqual(client.domain, "http://test-domain")
+
+        # Test with empty domain (should raise error)
+        with self.assertRaises(ValueError):
```
If you use the Metaclass, then you might need to update this to something like:
```python
        # Test with empty domain (should raise error)
        with self.assertRaises(TypeError) as e:
            AindMetadataServiceClient()
        self.assertIn("You must specify the server domain.", str(e.exception))
```

jtyoung84 reviewed (approved):
Thanks for the contribution! We can discuss ways to make the user experience better so they don't have to input the domain all the time. In the next major release, I plan to use OpenAPI to automatically generate the client code, which will make it easier to maintain.
